src/object/srv_obj.c

```diff
@@ -1713,29 +1715,80 @@ do_obj_ioc_begin(uint32_t rpc_map_ver, uuid_t pool_uuid,
 
 out:
 	dss_rpc_cntr_enter(DSS_RC_OBJ);
+	/** increment active request counter and start the chrono */
+	tls = obj_tls_get();
+	(void)d_tm_increment_gauge(&tls->ot_op_active[opc], 1, NULL);
+	ioc->ioc_start_time = daos_get_ntime();
 	ioc->ioc_began = 1;
 	return rc;
 }
 
+static inline unsigned int
+lat_bucket(uint64_t size)
+{
+	int nr;
+
+	if (size <= 256)
+		return 0;
+
+	/** return number of leading zero-bits */
+	nr =  __builtin_clzl(size - 1);
+
+	/** >4MB, return last bucket */
+	if (nr < 42)
+		return NR_LATENCY_BUCKETS - 1;
+
+	return 56 - nr;
+}
+
+static inline void
+obj_update_sensors(struct obj_io_context *ioc, int err)
+{
+	struct obj_tls		*tls = obj_tls_get();
+	struct d_tm_node_t	**lat;
+	uint32_t		opc = ioc->ioc_opc;
+	uint64_t		time;
+
+	(void)d_tm_decrement_gauge(&tls->ot_op_active[opc], 1, NULL);
+	(void)d_tm_increment_counter(&tls->ot_op_total[opc], 1, NULL);
+
+	if (unlikely(err != 0))
+		return;
+
+	/**
+	 * Measure latency of successful I/O only.
+	 * Use bit shift for performance and tolerate some inaccuracy.
+	 */
+	time = daos_get_ntime() - ioc->ioc_start_time;
+	time >>= 10;
+
+	switch (opc) {
+		case DAOS_OBJ_RPC_UPDATE:
+		case DAOS_OBJ_RPC_TGT_UPDATE:
+			(void)d_tm_increment_counter(&tls->ot_update_bytes,
+						     ioc->ioc_io_size, NULL);
+			lat = &tls->ot_update_lat[lat_bucket(ioc->ioc_io_size)];
```

Comment:
probably UPDATE and TGT_UPDATE should be separated? since UPDATE(time) == total of tgt_update(time).  Mixed them together might confuse metric information a bit.

Reply:
Let's discuss that on Skype.